tests/unit/s2n_get_protocol_version_test.c

```diff
@@ -0,0 +1,377 @@
+    }
+
+    /* Test protocol version getters on the server when a supported versions extension is received */
+    for (uint8_t server_version = S2N_TLS12; server_version <= S2N_TLS13; server_version++) {
+        for (uint8_t client_hello_version = S2N_SSLv3; client_hello_version <= S2N_TLS13; client_hello_version++) {
+            for (uint8_t client_supported_version = S2N_SSLv3; client_supported_version <= S2N_TLS13; client_supported_version++) {
+                DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+                EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+
+                if (server_version == S2N_TLS12) {
+                    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "test_all_tls12"));
+                } else {
+                    if (!s2n_is_tls13_fully_supported()) {
+                        continue;
+                    }
+                    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "test_all"));
+                }
+
+                uint8_t supported_versions_data[3] = { 0 };
+                struct s2n_blob supported_versions_blob = { 0 };
+                EXPECT_SUCCESS(s2n_blob_init(&supported_versions_blob, supported_versions_data, sizeof(supported_versions_data)));
+                EXPECT_OK(s2n_write_test_supported_versions_extension(&supported_versions_blob, client_supported_version));
+
+                /* The override_supported_versions client hello callback is used to overwrite the
+                 * supported versions extension before the extension is processed.
+                 */
+                struct s2n_override_extension_ctx context = {
+                    .extension_blob = supported_versions_blob
+                };
+                EXPECT_SUCCESS(s2n_config_set_client_hello_cb(config, s2n_override_supported_versions_cb, &context));
+
+                DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                        s2n_connection_ptr_free);
+                EXPECT_NOT_NULL(client);
+                EXPECT_SUCCESS(s2n_connection_set_config(client, config));
+
+                DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                        s2n_connection_ptr_free);
+                EXPECT_NOT_NULL(server);
+                EXPECT_SUCCESS(s2n_connection_set_config(server, config));
+
+                EXPECT_SUCCESS(s2n_client_hello_send(client));
+
+                /* Overwrite the client hello version according to the test case. */
+                struct s2n_stuffer *hello_stuffer = &client->handshake.io;
+                EXPECT_SUCCESS(s2n_stuffer_rewrite(hello_stuffer));
+                EXPECT_OK(s2n_write_protocol_version(hello_stuffer, client_hello_version));
+                EXPECT_SUCCESS(s2n_stuffer_write(&server->handshake.io, &hello_stuffer->blob));
```

Comment:
This test seems pretty complex. You overwrite the client hello version here, and overwrite the extension in the callback. Can you use the same method for both?

Reply:
Sorry for the complex tests. I tried to simplify them by moving more stuff into the client hello callback. I can't really test the same behavior of receiving an actual client hello with a modified version since the callback is invoked after the version is parsed. But I changed this to have the callback overwrite the version fields to fake having received a different client hello version.